File: openai-tutorials/5-Terminal_chat_with_a_customer_support_chatbot/tui_customer_support_chatbot.py
import os
import time
from dotenv import load_dotenv
from openai import OpenAI
client = OpenAI()
OpenAI.api_key = os.getenv('OPENAI_API_KEY')
from rich.console import Console
console = Console()

# Load environment variables from .env
load_dotenv()

# Existing file and assistant information
# Get file and assistant IDs from environment variables
file_id = os.getenv('OPENAI_FILE_ID')
assistant_id = os.getenv('OPENAI_ASSISTANT_ID')

# Create a new thread
my_thread = client.beta.threads.create()

# Loop until the user enters "quit"
while True:
    # Get user input
    user_input = input("User: ")

    # Check if the user wants to quit
    if user_input.lower() == "quit":
        console.print("\nAssistant: Have a nice day! :wave: \n", style="black on white")
        break

    # Add user message to the thread
    my_thread_message = client.beta.threads.messages.create(
        thread_id=my_thread.id,
        role="user",
        content=user_input,
        file_ids=[file_id]
    )

    # Run the assistant
    my_run = client.beta.threads.runs.create(
        thread_id=my_thread.id,
        assistant_id=assistant_id,
        instructions="Please address the user as Rok Benko."
    )

    # Initial delay before the first retrieval
    time.sleep(15)

    # Periodically retrieve the run to check its status
    while my_run.status in ["queued", "in_progress"]:
        keep_retrieving_run = client.beta.threads.runs.retrieve(
            thread_id=my_thread.id,
            run_id=my_run.id
        )

        if keep_retrieving_run.status == "completed":
            # Retrieve the messages added by the assistant to the thread
            all_messages = client.beta.threads.messages.list(
                thread_id=my_thread.id
            )

            # Display assistant message
            console.print(f"\nAssistant: {all_messages.data[0].content[0].text.value}\n", style="black on white")

            break
        elif keep_retrieving_run.status in ["queued", "in_progress"]:
            # Delay before the next retrieval attempt
            time.sleep(5)
            pass
        else:
            break

File: openai-tutorials/5-Terminal_chat_with_a_customer_support_chatbot/tui_customer_support_chatbot.js
const readline = require("readline");
const OpenAI = require("openai");
const openai = new OpenAI({
  apiKey: process.env.OPENAI_API_KEY,
});
const term = require("terminal-kit").terminal;

// Load environment variables from .env
require("dotenv").config();

// Existing file and assistant information
// Get file and assistant IDs from environment variables
const fileID = process.env.OPENAI_FILE_ID;
const assistantID = process.env.OPENAI_ASSISTANT_ID;

// Asynchronous function to create a new thread
async function createThread() {
  // Create a new thread
  const myThread = await openai.beta.threads.create();
  return myThread;
}

// Asynchronous function to get user input
function promptAsync(message) {
  // Creating a readline interface for reading lines from the standard input (keyboard)
  const rl = readline.createInterface({
    input: process.stdin, // Setting the input stream to the standard input (keyboard)
    output: process.stdout, // Setting the output stream to the standard output (console)
  });

  // Returning a Promise that resolves when the user enters something
  return new Promise((resolve) => {
    // Asking the user for input with the provided message
    rl.question(message, (userInput) => {
      // Closing the readline interface after receiving input
      rl.close();
      // Resolving the Promise with the user's input
      resolve(userInput);
    });
  });
}

// Function to simulate a delay using setTimeout and a Promise
function sleep(ms) {
  return new Promise((resolve) => setTimeout(resolve, ms));
}

// Asynchronous function to retrieve the run status and display the assistant's response
async function retrieveRun(threadId, runId) {
  while (true) {
    const keepRetrievingRun = await openai.beta.threads.runs.retrieve(
      (thread_id = threadId),
      (run_id = runId)
    );

    if (keepRetrievingRun.status === "completed") {
      // Retrieve the messages added by the assistant to the thread
      const allMessages = await openai.beta.threads.messages.list(
        (thread_id = threadId)
      );

      // Display assistant message
      term.inverse(
        "\nAssistant: ",
        allMessages.data[0].content[0].text.value,
        "\n"
      );

      break;
    } else if (
      keepRetrievingRun.status === "queued" ||
      keepRetrievingRun.status === "in_progress"
    ) {
      // Delay before the next retrieval attempt
      await sleep(5000);
    } else {
      break;
    }
  }
}

// Asynchronous function to execute the main logic
async function main() {
  // Create the thread and wait for it to be created
  const myThread = await createThread();

  let userInput;

  // Loop until the user enters "quit"
  while (true) {
    // Get user input
    console.log("");
    userInput = await promptAsync("User: ");
    console.log("");

    // Check if the user wants to quit
    if (userInput.toLowerCase() === "quit") {
      term.inverse("Assistant: Have a nice day!");
      console.log("\n");
      break;
    }

    // Add user message to the thread
    const myThreadMessage = await openai.beta.threads.messages.create(
      (thread_id = myThread.id),
      {
        role: "user",
        content: userInput,
        file_ids: [fileID],
      }
    );

    // Run the assistant
    const myRun = await openai.beta.threads.runs.create(
      (thread_id = myThread.id),
      {
        assistant_id: assistantID,
        instructions: "Please address the user as Rok Benko.",
      }
    );

    // Initial delay before the first retrieval
    await sleep(15000);

    // Periodically retrieve the run to check its status
    await retrieveRun(myThread.id, myRun.id);
  }
}

// Calling the main function to start the program
main();
